docs/client.rst

```diff
@@ -0,0 +1,137 @@
+
+Every week day at 10am (Pacific Time) we take part in a public daily stand-up,
+usually via a `meeting on Jitsi <https://meet.jit.si/QuickWizardsDanceHigh>`_
+(although the details of each daily meeting are published on the Gitter channel
+five minutes before the start of the meeting). All are welcome to contribute.
+
+Otherwise, read on.
+
+Client Architecture
+-------------------
+
+The SecureDrop client is a PyQt application. It's written using Python 3.5 and
+the Python bindings for the Qt UI framework (`PyQt <https://riverbankcomputing.com/software/pyqt/intro>`_).
+
+In the root directory of the repository are two important directories:
+``securedrop_client`` (containing the application code) and ``tests``
+containing our unit tests. You'll also find a Makefile in the root directory
+which defines commands to run commonly needed activities. Type, ``make`` to
+find out what commands are available.
+
+The code in the ``securedrop_client`` namespace is organised in the following
+way:
+
+* ``app.py`` - starts and configures the application.
+* ``logic.py`` - contains the application logic, encapsulated in the ``Client`` class.
+* ``models.py`` - holds all the `SQLAlchemy ORM model definitions <https://www.sqlalchemy.org/>`_ for interacting with the local Sqlite database.
+* ``storage.py`` - contains the functions needed for interacting with a remote SecureDrop API and the local database.
+* ``utils.py`` - generic utility functions needed throughout the application.
+* ``gui`` - this namespace contains two modules: ``main.py`` (containing the ``Window`` class through which all interactions with the user interface should happen) and ``widgets.py`` (containing all the custom widgets used by the ``Window`` class to draw the user interface).
+
+We try very hard to leep the application logic and UI code cleanly separated.
+Furthermore, we try equally hard to ensure the main GUI code always remains
+unblocked. For instance look at how the ``APICallRunner`` is used in
+``logic.py`` to make unblocked network calls to the remote API.
+
+We expect all classes, methods and functions to have docstrings describing the
+intention behind the code. We expect such docstrings to **remain up to date**
+as the code evolves. ;-)
+
+If possible, please use `Python type hints <https://docs.python.org/3.5/library/typing.html>`_
+for new code. We're going to transition the code base to this style in the
+not-too-distant future.
+
+Tests
+-----
+
+The files and directory structure found within the ``tests`` directory mirrors
+that of the files and directories in ``securedrop_client``. For instance, all
+the unit tests for the ``securedrop_client/logic.py`` module can be found in
+the ``tests/test_logic.py`` file.
+
+To run the complete test suite simply type::
+
+    make check
+
+Our code style checkers, full test suite and coverage checker will run and
+report any errors.
+
+We use the `PyTest testing framework <https://docs.pytest.org/en/latest/>`_ for
```

Comment:
We don't enforce this (docs/intent) among ourselves (well enough) and should clean this up before requiring new contributors do do the same. Further, sometimes a test is very obvious or commends scattered through are sufficient. However, I *would* support making this mandatory where each test has some sort of cucumber like specification in the docstring.